Memprof: provide the guarantee that an allocation callback is always run in the same thread the allocation takes place. Attempt II.

jhjourdan opened the pull request:
This PR is a second attempt at implementing the feature proposed by #9449.

In this new version, I use a separate extensible array for each thread, which contains the entries for the tracked blocks for which the allocation callback has not yet been called. Another shared array contains the entries for which the allocation callback has been called.

In addition, this PR contains a few code improvements which are related to this new data structure. In particular, tracking of young blocks has been made much simpler, since we know for sure that the allocating thread is the only one allowed to modify the local entry array (and hence no concurrent modification can happen when handling the different allocations of the same combined block).

This PR also gets rid of the `idx_ptr` pointers, which had relatively complex invariants. I am no completely sure that the replacement is better, though. Since this is in a separate commit, please tell me if we should of not use this.

Taking advantage of the removal of `idx_ptr`, this PR also lifts the restriction about `Thread.exit` in Memprof callbacks. Same remark as above: this is in a separate optional commit.

The diff of this PR is quite large, since it changes significantly the main data structure of Memprof. I tried to make independent commits to ease reviewing (only one commit should be non-trivial). Please review this one commit at a time.

Since this PR have more thread-local data structures, it should help rebasing on top of multicore. I'm happy to help.

Cc @stedolan, @gasche, @ctk21.

jhjourdan committed:
Memprof: refactor realloc_trackst.

jhjourdan committed:
Gc.Memprof.stop does no longer try to run pending callbacks.

This is only effetive in native mode, since function calls in bytecode
mode will trigger polling and hence run pending callbacks.

jhjourdan committed:
Memprof : get rid of the tracking_state struct, and create an entry_array struct, which only contains information specific to the array.

jhjourdan committed:
Whitespace, typos.

jhjourdan committed:
Memprof : Refactor [run_callback_exn] so that it also stores the value returned by the callback.

jhjourdan committed:
Memprof: provide the guarantee that an allocation callback is always run in the same thread the allocation takes place.

This is done by using a local entry array for each thread, containing
tracked blocks whose allocation callback has not yet been called.

This allows some simplification in the code running callbacks for
young allocations. Indeed, since the entry array is local to one
thread, we know for sure that it cannot be modified during a callback,
and therefore we no longer need to remember the indices of the
corresponding new entries.

jhjourdan committed:
Memprof: get rid of the idx_ptr pointers.

Instead, we use a thread-local variable [callback_status] which
contains the index of the corresponding entry when a callback is
running. We can do this since there can only be one running callback
at the same time in a given thread.

This lifts the restriction forbidding the call of Thread.exit from a
memprof callback.

jhjourdan committed:
Memprof/systhreads : call [caml_memprof_delete_th_ctx] when threads gets killed by fork.

jhjourdan commented:
One last comment: the first commit of this PR is actually #9653, which I'd loved to be merged first.

jhjourdan commented:
@stedolan : any chance you could tell me what you think about this? Of course, you don't need to do a thorough review, but I'd appreciate to know your opinion.

jhjourdan committed:
Changes.

jhjourdan committed:
Memprof: fix rounding error.

jhjourdan commented:
Would it be possible to get a review for this PR?

I know the diff is large, but I tried hard to split it into several commits, and this PR simplify some aspects of the implementation of Memprof.

I'd really like to see it merged for 4.12, and given that the diff is large, the review process may take some time. For the records, the first version of this PR, #9449, has been there since mid April, and did  not receive much attention since then..

xavierleroy commented:
@stedolan would you be available for reviewing this PR?  I'm afraid it will just drop into oblivion otherwise.

jhjourdan commented:
@stedolan, anychance you will have some time to have a look at this before 4.12?

stedolan commented:
Yes! Apologies for the slowness, it's on my list but I haven't found time yet. I'll have a look early next week.

jhjourdan commented:
Thanks !

And sorry if you feel a bit harassed : I am just afraid we won't have time before 4.12!

stedolan commented on otherlibs/systhreads/st_stubs.c:
```diff
@@ -431,20 +442,15 @@ static void caml_thread_remove_info(caml_thread_t th)
-    thr = next;
+  while (curr_thread->next != curr_thread) {
+    caml_memprof_stop_th_ctx(&curr_thread->next->memprof_ctx);
+    caml_thread_remove_info(curr_thread->next);
```
Is it a bug that this did not previously call `caml_thread_remove_info`?

jhjourdan replied:
Indeed, it seems to be a small memory leak.

stedolan replied:
Well spotted, then!

stedolan commented on runtime/memprof.c:
```diff
@@ -334,264 +346,322 @@ struct tracked {
+     the major heap ([young <= len]). */
+  uintnat young_idx;
+  /* There are no blocks to be deleted before this position
+     ([delete_idx <= entrieslen]). */
```
Should entrieslen be len in this comment?

stedolan commented on runtime/memprof.c:
```diff
@@ -334,264 +346,322 @@ struct tracked {
+  uintnat delete_idx;
+};
 
+#define MIN_ENTRIES_LOCAL_ALLOC_LEN 16
```
What's the reasoning behind having separate minimum entries lengths for local vs. global entry_array structures?

jhjourdan replied:
The local and global arrays do not store the same kind of data: the local array stores entries waiting for an alloc callback, so that it has a very small number of entries, and I don't expect that growing it is necessary in practice, while the global array contains all the rest, so that it may very well be much larger.

jhjourdan replied:
May I mark this as resolved?

stedolan commented:
This looks good! I'm very glad to see the back of idx_ptr. I'm going to have a more careful read of the reentrancy logic now, though, because I think that remains the trickiest part.

By the way, is there still a test of what happens when Thread.exit is called during a callback?

stedolan commented on runtime/memprof.c:
```diff
@@ -334,264 +346,322 @@ struct tracked {
+
+    if (t->deleted || t->callback_running) {
+      /* This entry is not ready. Ignore it. */
+      callback_idx++;
```
I think there's a small bug here. Suppose another thread is blocked in a promotion callback on callback_idx = 2. This code will then set callback_idx = 3. When the other thread finishes its promotion callback, it won't consider running the deallocation callback for callback_idx = 2, because the loop in caml_memprof_handle_postponed_exn will move on.

jhjourdan replied:
Thanks. There is a fix in 1749951 This is a bit hacky, since it involves checking for a special case in `run_callback_exn`. If you have a better idea, then I'm open, but I am afraid it involves adding quite a bit of complexity.

stedolan replied:
Thanks. I suspect there may be a simpler way to do this, I'll see if I can come up with one. Your fix looks correct, though.

stedolan commented on runtime/memprof.c:
```diff
@@ -334,264 +346,322 @@ struct tracked {
+  while (i < ea->len) {
+    if (!ea->t[i].deleted) {
+      if (ea->t[i].callback_running) {
+        if (callback_status == i) callback_status = j;
```
I think this is wrong if ea == entries_local. Can that happen?

jhjourdan replied:
Well spotted, fixed.

stedolan commented on runtime/memprof.c:
```diff
@@ -334,264 +346,322 @@ struct tracked {
-  t->idx_ptr = NULL;
+  /* The call above can move the tracked entry and thus invalidate
+     [t_idx] and [t]. */
+  if (ea == &entries_global) {
```
The following assertion here causes no failures in the testsuite:
```c
CAMLassert(t_idx == callback_status);
```
So I don't think the somewhat tricky invalidation/moving logic is actually tested anywhere.

stedolan replied:
See https://github.com/stedolan/ocaml/commit/0f864c94bd442ae3017a8f508daf27b10929d30d

stedolan commented:
While reading this I thought of a slightly simplification: since each `struct tracked` can only be executing a callback in one thread at a time, we can make the `struct tracked` point to the thread's `caml_memprof_th_ctx`, so that when updating indices in `flush_deleted` we don't need the callback/closure/iteration setup.

This is a bit easier to explain in code than text. See https://github.com/stedolan/ocaml/commit/25563fb04421408324aca8d20f5b675ac11672b6

jhjourdan committed:
Memprof: thread-local context is now stored as a pointer in systhread's data structures.

Reasons:
- Better abstraction in memprof.h
- Simpler Saving/restore functions
- We introduce a current's thread context, so that we don't need to do a spacial case for the current thread

jhjourdan commented:
> While reading this I thought of a slightly simplification: since each struct tracked can only be executing a callback in one thread at a time, we can make the struct tracked point to the thread's caml_memprof_th_ctx, so that when updating indices in flush_deleted we don't need the callback/closure/iteration setup.
>
> This is a bit easier to explain in code than text. See stedolan@25563fb

Thanks for the suggestion. I merged your ideas in the relevant commits of the PR.

stedolan committed:
Add a test for Memprof.stop within a Memprof callback

stedolan commented:
Thanks! I've written two more tests for the bits of this PR that I find trickiest. (see https://github.com/stedolan/ocaml/commits/memprof_thread_alloc_callback_2) One of them uncovers a segfault if we stop Memprof during a Memprof callback, mind having a look?

jhjourdan commented:
@stedolan : thanks for your review, I have included a test calling Thread.exit in a memprof callback. It is somewhat weak, since it does only a few allocations and hence does not check fully that the data structures stay in a valid state.

So now, I think this PR is read, since I have addressed all your comments. Shall we merge or ask an additional review?

jhjourdan commented:
(CI was broken, should be fixed now.)

stedolan commented:
> So now, I think this PR is read, since I have addressed all your comments. Shall we merge or ask an additional review?

I think you missed one comment: the test [here](https://github.com/stedolan/ocaml/commit/48928b999acbaaeb9ca62e35be6dc55f0b90a1cb) still causes a segfault.

jhjourdan commented:
> I think you missed one comment: the test here still causes a segfault.

Thanks. I have now fixed this.

The only issue with this PR is that the new test `moved_while_blocking.ml` fails with flambda because of #9978...

xavierleroy commented:
While the Flambda strangeness is being investigated, you can turn the test off in Flambda mode:
```
(* TEST
* hassysthreads
** no-flambda
include systhreads
*** bytecode
*** native
*)
```

jhjourdan commented:
> While the Flambda strangeness is being investigated, you can turn the test off in Flambda mode:

As you prefer. But I did not want to disable a test which reveals an actual bug! Actually it depends on whether #9978 can be investigated within short time bounds.

jhjourdan commented:
Thanks to the suggestion of @lthls, I fixed the test. CI should pass soon.

stedolan commented on runtime/memprof.c:
```diff
@@ -274,324 +378,293 @@ static value capture_callstack(int alloc_idx)
-    return Val_unit;
+  if (local->callback_status == CB_STOPPED) {
+    /* Make sure this entry has not been removed by [caml_memprof_stop] */
+    local->callback_status = CB_IDLE;
```
I don't understand why `local->callback_status` is changed here.

stedolan replied:
Never mind, I do now! I hadn't read the comment on CB_STOPPED properly: a thread is CB_STOPPED while it is executing a callback during which Memprof was stopped, and CB_IDLE once that callback has finished.

jhjourdan replied:
The documentation of `callback_status` says:

>   /* [callback_status] contains:
>     - CB_STOPPED if the current thread is running a callback, but
>       sampling has been stopped using [caml_memprof_stop];

Here, we go from the status "running a callback and stopped" to the status "not running a callback and stopped", hence `CB_STOPPED` is no longer applicable. This gets more coherent with the behavior of `caml_memprof_stop`, where we do:

```c
  if (ctx->callback_status != CB_IDLE) ctx->callback_status = CB_STOPPED;
```

I.e., we use `CB_STOPPED` only when a callback is currently running.

Moreover, not resetting `callback_status` would have the consequence that, at the next call to `caml_memprof_start`, `callback_status` would still contain `CB_STOPPED` (since `caml_memprof_start` does not touch this flag), which is clearly wrong. And no, you don't want `caml_memprof_start` to reset it, since that function could be called during a callback.

gasche commented:
The PR was approved and the CI is green. Should we merge the PR, or does @jhjourdan want to change the commit history?

(I looked at the history, my impression is that squashing everything would be not-nice, in particular it would make bisecting an issue within the PR much harder. There is a small amount of back-and-forth in the history, but I think it would be acceptable to merge right away if @jhjourdan doesn't happen to be a git history control addict.)

jhjourdan commented:
Ok, let me have a look at history then.

jhjourdan commented:
Alright, the history looks better now, I think. This is ready to merge and cherry-pick to 4.12. (I don't have the commit bit, though.)

jhjourdan commented:
Cherry-picked in 4.12 branch as part of commit 2c85e04ad1d25d30ae5366fb82c420e70dd95b07.
